### PR TITLE
Rewrite our test suite from scratch

### LINKED ISSUE #223: Rewrite our test suite from scratch
That issue may sound strange, but that is true.
Our comprehensive test suite (around 150 tests) is good, but at the same time not.
It's because it is based on assumptions, which is bad in general, and break all the tests when implementing unique id #222 
Current test suite has 2 assumptions:
- We assume that tests are run in the same order always, by filesystem ordering in directories, by the order defined in code inside files. It means that we can't split our tests into pieces and parallel execute it. Also it means that we are testing our use case, even though it covers many different parts, it might miss some. It is more like, functional or integrational tests, not unit tests. Currently we both simulate some common approach and unit test some functions
- Also, we assume that objects IDs are integers in range from 1 to N, which is also incorrect

Plus, this test suite was created in the early days of BitcartCC, a lot have changed, especially after our refactors (#217 and #219).

I propose to rewrite our test suite from scratch, no longer making any assumptions, making the test suite working for us long-term.

From discussion, we have decided that @xiaoxianma will initialize the initial structure of new tests, and I will join in. Any additional help welcome. We will re-think everything that was tested before and that should be tested now.

### PR DESCRIPTION
**Pull request by xiaoxianma**

### CODE REVIEW

File: requirements/test.txt [hunk 1]
```diff
@@ -2,6 +2,7 @@ pytest
 pytest-asyncio
 pytest-cov
 pytest-mock
+pytest-parametrization
```
Comment (MrNaif2018) [inline requirements/test.txt:5]: Reminder to self: re-run `scripts/compile-requirements.sh` (or you can do it, it requires python 3.7 installed)

File: tests/conftest.py [hunk 2]
```diff
@@ -6,8 +6,18 @@
 from starlette.testclient import TestClient
 
 from api import settings
+from api.db import db
 from main import app
 
+pytest_plugins = ["tests.fixtures.pytest.data"]
+
+
+@pytest.fixture(scope="class", autouse=True)
```
Comment (MrNaif2018) [inline tests/conftest.py:17]: How does `class` scope behave with regular functions? Will it create a new db for each function, but re-use database for a class of functions?

File: tests/fixtures/data/stores.json [hunk 3]
```diff
@@ -40,10 +40,10 @@
             "data": {
                 "name": "test5",
                 "wallets": [
-                    2
```
Comment (MrNaif2018) [inline tests/fixtures/data/stores.json:43]: Is this (and similar changes) because of database separation between functions, so that wallets created before aren't there? I guess we should add a test testing invoice creation in a store with multiple wallets connected just to be sure

File: tests/fixtures/pytest/data.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+@pytest.fixture(scope="class")
+async def user():
+    data = {
+        "email": f"superuser-{uuid4().hex[:8]}@example.com",
+        "password": "test12345",
+        "is_superuser": True,
+    }
+    user_obj = await create_user(**data)
+    return {**user_obj.to_dict(), **data}
```
Comment (MrNaif2018) [inline tests/fixtures/pytest/data.py]: For convenience, maybe this should return user model directly?

File: tests/fixtures/pytest/data.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+@pytest.fixture(scope="class")
+async def wallet(user):
+    return await create_wallet(user["id"])
+
+
+@pytest.fixture(scope="class")
+async def limited_user():
+    data = {
+        "email": f"nonsuperuser-{uuid4().hex[:8]}@example.com",
```
Comment (MrNaif2018) [inline tests/fixtures/pytest/data.py]: Why not using `unique_id` here and in the fixture above?

File: tests/helper.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+from api import models, utils
+from tests.fixtures import static_data
+
+
+async def create_user(**custom_attrs) -> models.User:
+    default_attrs = {
+        "email": f"user_{uuid4().hex[:8]}@gmail.com",
+        "password": "test12345",
+        "is_superuser": True,
+        "created": utils.time.now(),
```
Comment (MrNaif2018) [inline tests/helper.py]: What if we change the created field or others default values? This would need to be updated here too. I still think that we should use the http endpoints for that, or, at least, de-couple creation utils from views. Or an another way that we can use, we can also utilize pydantic schemes to fill in some defaults for us. As for some behaviour, like first user is superuser, we can just create an another user.

File: tests/helper.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    token_id = uuid4().hex
+    default_attrs = {
+        "id": token_id,
+        "user_id": user_id,
+        "app_id": "1",
+        "redirect_url": "test.com",
+        "permissions": ["full_control"],
+        "created": utils.time.now(),
+    }
+    return await create_model_obj(models.Token, default_attrs, custom_attrs)
```
Comment (MrNaif2018) [inline tests/helper.py]:
```suggestion
    return await create_model_obj(models.Token, **default_attrs, **custom_attrs)
```

To make consistent with other utils

Same in other utils

File: tests/helper.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+async def create_invoice(user_id: int, **custom_attrs) -> models.Invoice:
+    if "store_id" in custom_attrs:
+        store_id = custom_attrs.pop("store_id")
+    else:
+        store_id = (await create_store(user_id)).id
+    default_attrs = {
+        "price": random.randint(1, 10),
+        "currency": "USD",
+        "paid_currency": "BTC",
```
Comment (MrNaif2018) [inline tests/helper.py]: Note: do we create an already complete invoice by default, or maybe we should create an empty unpaid invoice? `paid_currency` is populated on payment only.

File: tests/helper.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    name = f"dummy_{uuid4()}"
+    if "store_id" in custom_attrs:
+        store_id = custom_attrs.pop("store_id")
+    else:
+        store_id = (await create_store(user_id)).id
+    default_attrs = {
+        "name": name,
+        "price": random.randint(1, 10),
+        "quantity": random.randint(100, 200),
+        "download_url": f"{name}.com",
```
Comment (MrNaif2018) [inline tests/helper.py]: Note: setting `download_url` affects email message generation. By default it is not set

File: tests/helper.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    return await create_model_obj(models.Product, default_attrs, custom_attrs)
+
+
+async def create_store(user_id: int, **custom_attrs) -> models.Store:
+    name = f"dummy_store_{uuid4()}"
+    default_attrs = {
+        "name": name,
+        "default_currency": "USD",
+        "email": f"{name}@gmail.com",
+        "email_host": "google.com",
```
Comment (MrNaif2018) [inline tests/helper.py]:
```suggestion
        "email_host": "smtp.gmail.com",
```

Or are those just dummy values? If so, there is no need to set them

File: tests/helper.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        "user_id": user_id,
+        "created": datetime.now(),
+    }
+    return await create_model_obj(models.Wallet, default_attrs, custom_attrs)
+
+
+async def create_store_wallet(user_id: int, custom_store_attrs: dict = {}, custom_wallet_attrs: dict = {}) -> dict:
+    store_obj = await create_store(user_id, **custom_store_attrs)
+    wallet_obj = await create_wallet(user_id, **custom_wallet_attrs)
+    await models.WalletxStore.create(wallet_id=wallet_obj.id, store_id=store_obj.id)
```
Comment (MrNaif2018) [inline tests/helper.py]: That's what I meant above. Managing m2m relationships and everything else is usually handled by the http routes.
By the way, isn't this record called twice because store's .create method should create an m2m relation by itself?

File: tests/helper.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+async def create_discount(user_id: int, **custom_attrs) -> models.Discount:
+    name = f"dummy_discount_{uuid4().hex}"
+    default_attrs = {
+        "user_id": user_id,
+        "name": name,
+        "percent": 5,
+        "description": "",
+        "promocode": "TEST",
+        "currencies": "",
+        "end_date": datetime.now(),
```
Comment (MrNaif2018) [inline tests/helper.py]: This means that this discount will be expired already, maybe you can add a timedelta of 1 day, for example

File: tests/helper.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        "created": datetime.now(),
```
Comment (MrNaif2018) [inline tests/helper.py]:
```suggestion
        "created": utils.time.now(),
```

Once again, I think that we shouldn't duplicate code this way

File: tests/helper.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+
+async def create_notification(user_id: int, **custom_attrs) -> models.Notification:
+    name = f"dummy_notf_{uuid4().hex}"
+    default_attrs = {
+        "user_id": user_id,
+        "name": name,
+        "provider": "NA",
+        "data": {},
+        "created": datetime.now(),
```
Comment (MrNaif2018) [inline tests/helper.py]: Same as above

File: tests/test_ext/test_export.py [hunk 15]
```diff
@@ -13,11 +14,12 @@ def test_merge_keys():
 
 
 @pytest.mark.asyncio
-async def test_db_to_json():
+async def test_invoice_db_to_json(user: dict):
+    await create_invoice(user_id=user["id"])
     items = await models.Invoice.query.gino.all()
     await utils.database.postprocess_func(items)
     json = list(db_to_json(items))
-    assert len(json) == 1
```
Comment (MrNaif2018) [inline tests/test_ext/test_export.py]: With db separation in place, shouldn't it return exactly 1 invoice?

File: tests/test_utils.py [hunk 16]
```diff
@@ -106,14 +108,16 @@ async def test_product_template(async_client, token):
         headers={"Authorization": f"Bearer {token}"},
     )
     assert resp.status_code == 200
+    store.user_id = resp.json()["user_id"]
```
Comment (MrNaif2018) [inline tests/test_utils.py]: What are those lines for?

File: tests/test_view/test_model_views.py [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+
+    def process_resp(self, resp, test, get_all=False):
+        to_check = self.status_mapping[test["status"]]
+        assert resp.status_code == to_check
+        if to_check == 200:
+            data = resp.json()
+            if get_all:
+                data = self.handle_get_all(data, test)
+            self.process_data(data)
+            if self.name == "users":
```
Comment (MrNaif2018) [inline tests/test_view/test_model_views.py]: What is this for?

File: tests/test_view/test_model_views.py [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+    tests = json_module.loads(open("tests/fixtures/data/wallets.json").read())
+
+
+class TestStores(ViewTestMixin):
+    name = "stores"
+    auth = True
+    tests = json_module.loads(open("tests/fixtures/data/stores.json").read())
+
+    @pytest.fixture(scope="class", autouse=True)
+    def setup(self, wallet, notification):
```
Comment (MrNaif2018) [inline tests/test_view/test_model_views.py]: What is this for?
Also, is `data_update` hook reserved for the future? What is it intended to do? Load related?

File: tests/test_view/test_views.py [hunk 19]
```diff
@@ -231,33 +43,34 @@ def test_rate(client: TestClient):
     assert client.get("/cryptos/rate?fiat_currency=test").status_code == 422
 
 
-def test_wallet_history(client: TestClient, token: str):
+def test_wallet_history(client: TestClient, token: str, wallet):
     headers = {"Authorization": f"Bearer {token}"}
-    assert client.get("/wallets/history/1", headers=headers).status_code == 404
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: This test now doesn't test non-existing wallets and other edge cases. Is it intentional?

File: tests/test_view/test_views.py [hunk 20]
```diff
@@ -273,26 +86,26 @@ def test_noauth(client: TestClient):
...
     resp = client.get("/users/me", headers={"Authorization": f"Bearer {token}"})
     assert resp.status_code == 200
     j = resp.json()
-    assert j.items() > {"is_superuser": True, "id": 1, "email": "testauth@example.com"}.items()
+    assert j.items() > {"is_superuser": True, "email": user["email"]}.items()
     assert "created" in j
 
 
-def test_wallets_balance(client: TestClient, token: str):
+def test_wallets_balance(client: TestClient, token: str, wallet):
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: Unused param?

File: tests/test_view/test_views.py [hunk 21]
```diff
@@ -350,39 +163,56 @@ async def check_ws_response2(ws):
...
     }
 
 
-def test_categories(client: TestClient):
-    assert client.get("/products/categories").status_code == 422
-    resp = client.get("/products/categories?store=1")
-    resp2 = client.get("/products/categories?store=2")
+@pytest.mark.asyncio
+@Parametrization.autodetect_parameters()
+@Parametrization.case(name="single", categories=["all"])
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: Note: as far as I see, this test would create a product with category `all`. The intention of first test was to ensure that `all` category is returned in any case, even when no categories are available

File: tests/test_view/test_views.py [hunk 22]
```diff
@@ -350,39 +163,56 @@ async def check_ws_response2(ws):
...
+        await create_product(user["id"], store_id=store_id, category=category)
+    resp = await async_client.get(f"/products/categories?store={store_id}")
     assert resp.status_code == 200
-    assert resp.json() == ["all"]
-    assert resp2.status_code == 200
-    assert resp2.json() == ["all", "Test"]
+    assert resp.json() == categories
 
 
 def check_token(result):
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: Note: it is now unused. Can either be removed or just the checking code can move here.

I think we should add a flake8 plugin to check for dead code automatically

File: tests/test_view/test_views.py [hunk 23]
```diff
@@ -393,63 +223,97 @@ def check_token(result):
...
-    assert len(result) == 1
-    result = result[0]
-    check_token(result)
-
-
-def test_token_current(client: TestClient, token: str):
+@Parametrization.autodetect_parameters()
+@Parametrization.case(name="unauthorized", authorized=False)
+@Parametrization.case(name="authorized", authorized=True)
+def test_token(client: TestClient, token_obj, authorized: bool):
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: Hmm, in some cases parametrization adds more complexity than needed. Maybe we could stick to non-parametrized tests for those? It's creating unnecessary branches while it just checks unauthorized access.

File: tests/test_view/test_views.py [hunk 24]
```diff
@@ -393,63 +223,97 @@ def check_token(result):
...
+    else:
+        assert resp.status_code == 401
+
+
+def test_token_current(client: TestClient, token: str, user: dict):
     assert client.get("/token/current").status_code == 401
     resp = client.get("/token/current", headers={"Authorization": f"Bearer {token}"})
     assert resp.status_code == 200
     j = resp.json()
-    check_token(j)
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: check_token can be used there without any issues

File: tests/test_view/test_views.py [hunk 25]
```diff
@@ -393,63 +223,97 @@ def check_token(result):
...
+    assert isinstance(j, dict)
+    assert j["user_id"] == user["id"]
+    assert j["app_id"] == "1"
+    assert j["redirect_url"] == "test.com"
+    assert j["permissions"] == ["full_control"]
     assert j["id"] == token
 
 
-def test_token_count(client: TestClient, token: str):
-    assert client.get("/token/count").status_code == 401
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: Same here. Maybe for such cases like that we should avoid parametrization

File: tests/test_view/test_views.py [hunk 26]
```diff
@@ -393,63 +223,97 @@ def check_token(result):
...
+    if authorized:
+        if user_exists:
+            resp.status_code == 200
+        else:
+            resp.status_code == 404
+    else:
+        resp.status_code == 401
 
 
-def test_create_tokens(client: TestClient, token: str):
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: See, the test is waay much simpler. We should use those

File: tests/test_view/test_views.py [hunk 27]
```diff
@@ -460,51 +324,62 @@ def test_create_tokens(client: TestClient, token: str):
...
+        json={"email": limited_user["email"], "password": limited_user["password"], "permissions": ["server_management"]},
+    )
+    assert resp.status_code == 422
     # Non-strict mode: silently removes server_management permission
     resp = client.post(
         "/token",
-        json={**LIMITED_USER_DATA, "permissions": ["server_management"], "strict": False},
+        json={
+            "email": limited_user["email"],
+            "password": limited_user["password"],
```
Comment (MrNaif2018) [inline tests/test_view/test_views.py]: As for this, we can just move password to constants and avoid exposing it to outside

File: tests/test_view/test_views_paginations.py [hunk 28]
```diff
@@ -0,0 +1,59 @@
+import pytest
```
Comment (MrNaif2018) [inline tests/test_view/test_views_paginations.py]: I think the directory should be renamed to `test_views`

File: tests/fixtures/data/invoices.json [hunk 29]
```diff
@@ -144,19 +144,19 @@
             "data": {
                 "price": 0.2,
                 "products": [
-                    2
+                    1
```
Comment (MrNaif2018) [inline tests/fixtures/data/invoices.json:147]: By the way: does it fix the issues we had, relying on object id? Probably `ViewTestMixin` needs a bigger refactor? As for now, it still relies on object id.

**Review by MrNaif2018 — CHANGES_REQUESTED** [review]
Hi! Thank you very much for your PR! Please check the review comments

File: tests/conftest.py (hunk 2 above)
Comment (xiaoxianma) [reply tests/conftest.py:17]: if tests exist in class level, db will be only invoked once like test_model_views
if there are no class of testing, then db will be invoked of each function

Comment (MrNaif2018) [reply tests/conftest.py:17]: Ok, then it's exactly what we need! Great!

File: tests/fixtures/data/stores.json (hunk 3 above)
Comment (xiaoxianma) [reply tests/fixtures/data/stores.json:43]: I think this is wallet id instead of count

File: tests/fixtures/pytest/data.py (hunk 4 above)
Comment (xiaoxianma) [reply tests/fixtures/pytest/data.py]: user model won't include password field, which is needed to create token (token testing)

File: tests/fixtures/pytest/data.py (hunk 5 above)
Comment (xiaoxianma) [reply tests/fixtures/pytest/data.py]: sure. I will make a change to use unique_id

File: tests/fixtures/data/invoices.json (hunk 29 above)
Comment (xiaoxianma) [reply tests/fixtures/data/invoices.json:147]: It is still relying on object id and it should rely on due to some tests required to retrieve obj by id (now integer id, future unique id)

File: tests/fixtures/data/stores.json (hunk 3 above)
Comment (xiaoxianma) [reply tests/fixtures/data/stores.json:43]: wallet id 2 means previous wallet id 1 is deleted by previous tests

**Review by MrNaif2018 — APPROVED** [review]
Thank you very much for your PR! I will apply some last fixes and re-test it and merge it then
